File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.js
/**
 * @param {character[]} letters
 * @param {character} target
 * @return {character}
 */
const nextGreatestLetter = (letters, target) => {
    if (letters[0] > target || target >= letters[letters.length - 1])
    return letters[0];
    
    let lo = 0, hi = letters.length - 1;
    while (lo < hi) {
        let mid = lo + Math.floor((hi - lo) / 2);
        if (letters[mid] <= target) {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    return letters[lo];
};

File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        left, right, letter = 0, len(letters) - 1, letters[0]
        while left <= right:
            mid = left + (right - left) // 2
            if target < letters[mid]:
                letter = letters[mid]
                right = mid - 1
            else:
                left = mid + 1
        return letter
                
                